sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -118,9 +128,18 @@ def _csr_polynomial_expansion(ndarray[DATA_T, ndim=1] data,
     cdef ndarray[INDEX_T, ndim=1] expanded_indptr = ndarray(
         shape=num_rows + 1, dtype=indptr.dtype)
 
+    # if the expanded_dimensionality is too big, we need to cast
+    # indices to int64. If total_nnz is too big, we need to cast
+    # expanded_indptr to int64 too.
+    # TODO: check that the casting is indeed done
+    if expanded_dimensionality > MAX_INT32:
+        indices = np.array(indices, dtype=np.int64)
```

Comment:
If `indices` is of dtype `np.int32`, it would be strictly typed to `indices` at this line. This means that we can not have it point to a int64 array.

I think we need to factor out the code that does the expansion into it's own fused function. The original `_csr_polynomial_expansion` will decide if it should cast `indices` and friends.